### PR TITLE
Test with JDK 17. Update Mockito and JUnit 5.x

### PR DESCRIPTION
**Pull request by martin-g**
Continuation of #407 
I will rebase it to `master` once #407 is merged!

### CODE REVIEW

**Review by eolivelli — APPROVED** [review]
Lgtm

@nicoloboschi you are working on jdk17 in Pulsar and BookKeeper 
Do you want to take a look?

**Comment by martin-g** [comment]
https://github.com/apache/curator/pull/410/commits/f6bac3372130882ee1991b087283842f542474ad "fixes" the problem in org.apache.curator.framework.imps.GzipCompressionProvider but it does not look nice.
I'd recommend a deeper look here!

File: curator-framework/src/main/java/org/apache/curator/framework/imps/GzipCompressionProvider.java [hunk 1]
```diff
@@ -45,6 +45,23 @@
...
+            // e.g. 18-ea, take 18
+            JAVA_VERSION = Integer.parseInt(JAVA_VERSION_STR.split("-", 2)[0]);
+        } else if (JAVA_VERSION_STR.contains(".")) {
+            // e.g. 17.0.1, take 17
+            JAVA_VERSION = Integer.parseInt(JAVA_VERSION_STR.split("\\.", 2)[0]);
+        } else {
+            JAVA_VERSION = Integer.parseInt(JAVA_VERSION_STR, 10);
+        }
+    }
+    private static final byte OS_BIT = JAVA_VERSION >= 16 ? (byte) -1 : 0;
```
Comment (nicoloboschi) [inline curator-framework/src/main/java/org/apache/curator/framework/imps/GzipCompressionProvider.java:63]: could you add a comment explaining why this is needed ?

**Review by nicoloboschi — CHANGES_REQUESTED** [review]

File: curator-framework/src/main/java/org/apache/curator/framework/imps/GzipCompressionProvider.java (hunk 1 above)
Comment (martin-g) [reply curator-framework/src/main/java/org/apache/curator/framework/imps/GzipCompressionProvider.java:63]: I just reverse engineered it.
I don't know what exactly has changed in JDK 16 that caused the breakage here.
This is the reason I said `I'd recommend a deeper look here!`. Someone who knows this code better may have better idea what is going on.

Comment (nicoloboschi) [reply curator-framework/src/main/java/org/apache/curator/framework/imps/GzipCompressionProvider.java:63]: I think this is the related issue 
https://bugs.openjdk.java.net/browse/JDK-8244706

Comment (martin-g) [reply curator-framework/src/main/java/org/apache/curator/framework/imps/GzipCompressionProvider.java:63]: I will add it as a comment!

**Comment by martin-g** [comment]
Now the build on JDK 17 fails due to Jersey.
https://github.com/apache/curator/blob/fe50da4904aa4ca4f51b473ad298ae664358ff3c/pom.xml#L84-L86 says that it is not worth it to upgrade.

**Comment by eolivelli** [comment]
I am +1 to upgrading Jersey.
That comment (https://stackoverflow.com/questions/17098341#22033825) is very old.

Also I am not sure about why we are using Jersey in Curator...
do you want to try to upgrade please ?

**Comment by martin-g** [comment]
> Also I am not sure about why we are using Jersey in Curator...
> do you want to try to upgrade please ?

It is used only in curator-x-discovery-server/src/test/java/org/apache/curator/x/discovery/server/jetty_jersey/TestMapsWithJersey.java

Is it valueable ? Should I try to upgrade it or just delete it ?

**Comment by eolivelli** [comment]
> Is it valueable ? Should I try to upgrade it or just delete it ?

if we can keep the existing tests it is better, and it would be great to convert the test

that said, I am not sure about the real value of that test
maybe @Randgalt or @cammckenzie  know better than me.

**Comment by martin-g** [comment]
I've updated Jersey to it latest 2.x version but there is still some problem - the MapDiscoveryContext is not seen as a `@Provider` and `org.apache.curator.x.discovery.server.jetty_jersey.MapDiscoveryResource`'s constructor fails with a NPE. I'll try to debug it later.

**Comment by tisonkun** [comment]
@martin-g code conflict. If you'd like to move forward this effort, please rebase the commits so that I can give a review.

**Review by eolivelli — APPROVED** [review]
LGTM

**Comment by tisonkun** [comment]
It seems tests failed steadily.

**Pull request closed without merging**